binary_search_tree/tree.py

```diff
@@ -14,43 +14,128 @@ class Tree:
+            self.root = TreeNode(key, value) 
+        else:
+            self.add_helper(self.root, key, value)
 
-    # Time Complexity: 
-    # Space Complexity: 
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(1)
     def find(self, key):
-        pass
+        if self.root == None:
+            return None
+        
+        current = self.root
+        while current !=  None:
+            if current.key == key:
+                return current.value
+            elif current.key < key:
+                current = current.right
+            else:
+                current = current.left
+            
+        return None
+        
+
+    def inorder_helper(self, current, trav_list):
+        if current == None:
+            return 
+
+        self.inorder_helper(current.left, trav_list)
+        trav_list.append({"key":current.key, "value":current.value})
+        self.inorder_helper(current.right, trav_list)
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def inorder(self):
-        pass
+        trav_list = []
+        self.inorder_helper(self.root, trav_list)
+        return trav_list
+
+
+    def preorder_helper(self, current, trav_list):
+        if current == None:
+            return 
 
-    # Time Complexity: 
-    # Space Complexity:     
+        trav_list.append({"key":current.key, "value":current.value})
+        self.preorder_helper(current.left, trav_list)
+        self.preorder_helper(current.right, trav_list)
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)    
     def preorder(self):
-        pass
+        trav_list = []
+        self.preorder_helper(self.root, trav_list)
+        return trav_list
+
+
+    def postorder_helper(self, current, trav_list):
+        if current == None:
+            return
 
-    # Time Complexity: 
-    # Space Complexity:     
+        self.postorder_helper(current.left, trav_list)
+        self.postorder_helper(current.right, trav_list)
+        trav_list.append({"key":current.key, "value": current.value})
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)    
     def postorder(self):
-        pass
+        trav_list = []
+        self.postorder_helper(self.root, trav_list)
+        return trav_list
+
 
-    # Time Complexity: 
-    # Space Complexity:     
+    def height_helper(self, current):
+        if current == None:
+            return 0
+
+        left_height = self.height_helper(current.left)
+        right_height = self.height_helper(current.right)
+
+        if left_height > right_height:
+            return left_height + 1
+        return right_height + 1
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(log n)    
     def height(self):
-        pass
+        return self.height_helper(self.root)
+        
 
 
 #   # Optional Method
-#   # Time Complexity: 
-#   # Space Complexity: 
+#   # Time Complexity: O(n)
+#   # Space Complexity: O(n)
     def bfs(self):
```

Comment:
👍 , but the space complexity is O(h) or O(log n) for a balanced tree because the call stack never gets larger than the height of the tree.